### ISSUE TITLE
Getting Typescript compilation error when including src/shortcuts.ts

### ISSUE THREAD

**Issue by sahil87** [jawj/zapatos#31]
I am getting the following error while trying to include the src/shortcuts.ts file.

```bash
Using ts-node version 8.10.2, typescript version 3.9.7
Compilation error in /src/dev/zapatos/src/shortcuts.ts
[ERROR] 20:51:48 ⨯ Unable to compile TypeScript:
dev/zapatos/src/shortcuts.ts:347:5 - error TS7005: Variable 'tableAliasSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

347     tableAliasSQL = aliasedTable === table ? [] : sql<string>` AS ${aliasedTable}`,
        ~~~~~~~~~~~~~
dev/zapatos/src/shortcuts.ts:353:5 - error TS7005: Variable 'colsLateralSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

353     colsLateralSQL = lateralOpt === undefined ? [] :
        ~~~~~~~~~~~~~~
dev/zapatos/src/shortcuts.ts:356:5 - error TS7005: Variable 'colsExtraSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

356     colsExtraSQL = extrasOpt === undefined ? [] :
        ~~~~~~~~~~~~
dev/zapatos/src/shortcuts.ts:360:5 - error TS7005: Variable 'whereSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

360     whereSQL = where === all ? [] : sql` WHERE ${where}`,
        ~~~~~~~~
dev/zapatos/src/shortcuts.ts:361:5 - error TS7005: Variable 'orderSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

361     orderSQL = !allOptions.order ? [] :
        ~~~~~~~~
dev/zapatos/src/shortcuts.ts:367:5 - error TS7005: Variable 'limitSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

367     limitSQL = allOptions.limit === undefined ? [] : sql` LIMIT ${param(allOptions.limit)}`,
        ~~~~~~~~
dev/zapatos/src/shortcuts.ts:368:5 - error TS7005: Variable 'offsetSQL' implicitly has an 'any[] | SQLFragment<any[]>' type.

368     offsetSQL = allOptions.offset === undefined ? [] : sql` OFFSET ${param(allOptions.offset)}`,
        ~~~~~~~~~
```

Am using typescript 3.9.7.
Here's my tsconfig.json:
```json
{
  "compilerOptions": {
    "module": "CommonJS",
    "noImplicitAny": true,
    "preserveConstEnums": true,
    "lib": ["ES2020", "DOM"],
    "target": "ES2020",
    "esModuleInterop": true,
    "jsx": "preserve",
    "typeRoots": ["node_modules/@types", "src/@types"],
  },
  "include": [.......],
  "exclude": []
}
```

If there's any more info you like to know let me know.

**Comment by jawj** [comment]
Hmm, I've not seen this before. Are you importing from the local copy created when you run `npx zapatos`, and do you still have the problem when you import all of `./zapatos/src` or `./zapatos/src/index` rather than only `shortcuts.ts`?

If so, it would help me investigate if you could share a minimal test case (e.g. as a new small repo).

**Comment by sahil87** [comment]
Found it. `"strict": true` needs to be added to tsconfig.json to remove the error. It's usually there in the default tsconfig.json - I had manually removed it.

Anyway, here's the example repo that replicates the issue: https://github.com/sahil87/zapatos-example

After adding it back to my tsconfig, everything works now. To be more specific, "strictNullChecks" is needed.
Sorry for the trouble @jawj , closing this.

**Comment by jawj** [comment]
Hmm, this is actually pretty weird. I just tried turning off `strict` in one of my projects, and it breaks type inference for `select` queries with `lateral` (the extra column(s) introduced by the join are just lost). Seems like it breaks a bunch of other things too, in a way I find quite surprising. I'll try to investigate this further at some point ...

**Comment by jawj** [comment]
Note to self: here's [a one-line test-case](https://www.typescriptlang.org/play?strictNullChecks=false&ts=4.0.0-beta#code/FAYw9gdgzgLgBADzgXjgWQIYwBYDoBOGEAJmALYAUAlHADxwAMuArHAPxwDaAunAFxwAjAG5gwIA).

The error can be fixed either by turning `strictNullChecks` back on or by changing `[]` to `<const>[]`.

**Comment by jawj** [comment]
Best I can tell, this is [expected TS behaviour](https://github.com/microsoft/TypeScript/issues/13140), so it's down to Zapatos to deal with `strictNullChecks` being turned off in host projects.

**Comment by jawj** [comment]
I've now released 0.1.57, which should fix the problems in `shortcuts.ts` when `strictNullChecks` are off.

But I've also added notes to the docs that `strictNullChecks` are required in order for the columns, lateral and extras options to the select shortcuts to function.

See also #38.
